### ISSUE TITLE
Matching Networks for One Shot Learning

### ISSUE THREAD

**Issue by sweaterr** [sweaterr/readinglist#304]
https://arxiv.org/pdf/1606.04080.pdf

**Comment by sweaterr** [comment]
“one-shot” learning
* learning a class from a single labelled example

support set S: 데이터셋
$C_S$: 분류기
$S => C_S(.)$

### Model Architecture

a neural attention mechanism

Matching Networks
* produce sensible test labels for unobserved classes without any changes to the network
* $S=\\{\\}$

**Comment by sweaterr** [comment]
[Matching networks for one shot learning](https://blog.acolyer.org/2017/01/03/matching-networks-for-one-shot-learning)
* something like k-nearest neighbours (KNN) doesn’t require any training at all
* these models there is some external memory and an attention mechanism which is used to access the memory. 
* It’s a network that learns how to learn a classifier from only a very small number of examples.
* $\hat{y}=\sum_{i=1}^{k}{a(\hat(x),x_i)y_i}$
  * attention mechanism $a$는 $x_i$와 $\hat{x_i}$의 유사도를 정해주는 역할

**Comment by sweaterr** [comment]
![image](https://user-images.githubusercontent.com/1518919/36956016-30183df6-206f-11e8-9141-bc0458d9df4b.png)

**Comment by sweaterr** [comment]
## 2.1.2 Full Context Embeddings 
* 외부 메모리가 있는 뉴럴넷을 원-샷 학습에 사용
* 임베딩 함수 $f$와 $g$는 이미지를 $\hat{y}=\sum_{i=1}^{k}{a(\hat(x),x_i)y_i}$의 분류 함수의 최대 정확도를 얻는 피쳐 공간으로 가게 한다.
* S를 모두 조건부로 사용하는 분류 전략 $P(.|\hat{x},S)$을 사용하지만, S의 각 이미지를 서로 독립적으로 임베딩하는 것은 약간 naive 하다. 더 나아가 S는 어떻게 f를 통해 이미지 $\hat{x}$을 임베딩하는지 수정할 수 있어야 한다.
* $g(x_i)$ => $g(x_i,S)$, S 전체를 보고, $x_i$를 어떻게 임베딩할지 결정한다. $x_i$와 $x_j$가 매우 가까울 때 유용할 수 있다. S가 주어졌을때, $x_i$를 어떻게 인코딩하는지에 대한 문제는 LSTM으로 풀 수 있다.

**Comment by sweaterr** [comment]
### A.1 The Fully Conditional Embedding f
* example $\hat{x}$, batch $B$
* $f(\hat{x},S) = \text{attLSTM}(f'(\hat{x}), g(S), K)$
  * 테스트셋 $\hat{x}$에 대한 임베딩은, S를 이용해서 구하는데, 테스트셋의 임베딩 피쳐 f'와 S의 임베딩 피쳐, 처리 횟수 K를 이용한다.

**Comment by sweaterr** [comment]
## 2.2 Training Strategy
* $S → c(\hat{x})$: S를 분류 함수에 매핑
*  훈련 과정은 테스트 시간에 추론과 매칭하기 위해  조심스레 선택되어야 한다.
* 훈련 중에 보지 못한 클래스가 포함된 support set S'로도 잘 동작하여야 한다
* $L$ = 가능한 레이블 집합
* $T$ = 가능한 레이블 집합에 대한 분포
* $L \sim T$ =

**Comment by sweaterr** [comment]
## Related Work

3.1 Memory Augmented Neural Networks
* Neural Turing Machine 
* Memory Networks

3.2 Metric Learning
*  Neighborhood Component Analysis

**Comment by sweaterr** [comment]
## 4 Experiments
* N-way k-shot 학습 과제
* $L'$ = one-shot 테스트용으로 빼논 데이터
* $\neq L'$ = 트레이닝 데이터
* Baseline
  * [MANN](http://proceedings.mlr.press/v48/santoro16.pdf) 
  * [Convolutional Siamese Net](https://www.cs.cmu.edu/~rsalakhu/papers/oneshot1.pdf)

* ImageNet
  *

**Comment by sweaterr** [comment]
모델의 복잡함에 비해, 성능향상 정도가 낮은 것 같다.

**Comment by sweaterr** [comment]
## 트레이닝 프로토콜
* "one-shot learning is much easier if you train the network to do one-shot learning"
* 데이터셋에서 훈련 " episode"를 다음과 같이 생성한다.
  1.  훈련셋에서 task T를 샘플링 한다. 예를 들어, 5개의 레이블을 선택하고, 레이블마다 5개까지의 예제를 뽑는다(5-25개의 예제가 있다)
  2. 하나의 에피소드 샘플을 생성하기 위해서,  레이블셋 L을 샘플링한다(예를 들어 {개,고양이}). 그 후 지원 셋 S와 로스를 계산할 배치 B를 샘플링하기 위해 L을 사용한다.

**Comment by sweaterr** [comment]
## The model

I find the paper's model description slightly wordy and unclear, but basically we're building a differentiable nearest neighbor++. The output $\hat{y}$ for a test example $\hat{x}$ is computed very similar to what you might see in Nearest Neighbors:

![image](https://user-images.githubusercontent.com/1518919/36975996-524b576e-20bf-11e8-996f-2aca64081a07.png)

where a acts as a kernel, computing the extent to which \hat{x} is similar to a training example x_i, and then the labels from the training examples (y_i) are weight-blended together accordingly. The paper doesn't mention this but I assume for classification y_i would presumbly be one-hot vectors.

Now, we're going to embed both the training examples x_i and the test example \hat{x}, and we'll interpret their inner products (or here a cosine similarity) as the "match", and pass that through a softmax to get normalized mixing weights so they add up to 1. No surprises here, this is quite natural:

**Comment by sweaterr** [comment]
**Embedding the training examples.**  This (the function g) is a bidirectional LSTM over the examples:

![image](https://user-images.githubusercontent.com/1518919/36976226-2c71dfda-20c0-11e8-9e70-72a9ba940a29.png)

즉 i 번째 예제 $ x_i $의 인코딩은 양방향 네트워크의 숨겨진 상태를 통해 전 된 "원시"embedding $ g '(x_i) $와 그 친구의 임베딩의 함수입니다. 즉, 각각의 트레이닝 예제는 그 자체의 뿐만 아니라 그 집합 안 모든 친구들의 함수이다. 일반적인 최근접 이웃에서는 훈련 세트의 다른 데이터 포인트의 함수로 예제의 표현을 변경하지 않기 때문에 위의 ++의 일부입니다.

**Comment by sweaterr** [comment]
#### Embedding the test example. 

이(f 함수)는 고정된 양 (K 타임 스텝)을 처리하는 LSTM이며  and at each point also attends over the examples in the training set..

인코딩은 LSTM의 마지막 숨겨진 상태입니다.

다시 말하지만, 우리는 네트워크가 훈련 예제에 따라 테스트 예제의 인코딩을 변경하도록 허용하고 있습니다.

![image](https://user-images.githubusercontent.com/1518919/36977705-6a8183c0-20c5-11e8-85fc-0cbbdb092f32.png)

이것은 처음에는 무서워 보이지만 실제로는 바닐라 LSTM이고 각 타임 스텝의 입력은 일정합니다 ($f '(\hat {x}$, an encoding of the test example all by itself) 숨겨진 상태는 이전 숨겨진 상태의 함수일뿐만 아니라 concat된 readout vector r의 함수이다. r은 (위에서 g를 통해 인코딩된) 훈련 예제에 대해서 어텐딩함으로써 얻을 수 있다.

**Comment by sweaterr** [comment]
## Experiments

Omniglot [14]은 각각 20 개의 예를 가진 1623 자의 MNIST와 같은 낙서 데이터 세트이다.

이미지 인코더는 [3x3 CONV 64 필터, batchnorm, ReLU, 2x2 max pool]의 4 개 모듈로 구성된 CNN입니다.

원본 이미지는 원본 28x28에서 1x1x64로 크기가 조정되므로 2 다운 샘플링을 4번하면 비율은 1/16, 28/16은 1보다 큰 정수가 아니므로, 무엇인가 이상합니다.

VALID convs를 사용한 것일까요?

![image](https://user-images.githubusercontent.com/1518919/36979133-e5b1498c-20c9-11e8-86aa-65e159924633.png)

매칭넷이 제일 좋습니다.

FCE (Full Conditional Embedding)은 Section 2.1.2의 "Full Context Embeddings"대신 여기에 사용되지 않고 훨씬 더 잘 작동하지 않는다고 말합니다.

Finetuning은 베이스 라인에서 조금은 도움이 되지만 매칭넷으로는 도움이 되지 않습니다(이상함).

이 표의 비교는 다소 혼란 스럽습니다. 

* MANN 숫자 82.8 %와 94.9 %를 그들의 논문에서 찾을 수 없습니다 [21]. 그들이 어디에서 왔는지 분명하지 않습니다.
* 나는 또한 Siamese Net [11] 논문에서이 수치를 발견 할 수 없다.
* 내가 이것을 정확히 이해했다면, 그들은 1- 샷 20- 웨이에서 95.2 %를, 매칭 네트는 93.8 %를 보여 주며, 인간은 95.5 %로 추정합니다. That is, the results here appear weaker than those of Lake et al,, but one should keep in mind that the method here is significantly more generic and does not make any assumptions about the existence of strokes, etc., and it's a simple, single fully-differentiable blob of neural stuff.

**Comment by sweaterr** [comment]
## Conclusions
이 좋은 논문은 효과적으로 종단 간 훈련을 이룰 수 있는 가장 가까운 이웃을 개발합니다.

A few concerns:

#### 1

쌍방향 LSTM (순서 불변 계산이 아님)이 일련의 학습 예제를 인코딩합니다.

저자는 실제로 사용된 순서에 대해 이야기하지는 않습니다. 아마 무작위일 것이라고 예상합니다. 그렇지 않다면 언급했어야 합니다.

This can be solved by using a recurrent attentional mechanism instead, as the authors are certainly aware of and as has been discussed at length in ORDER MATTERS: SEQUENCE TO SEQUENCE FOR SETS, where Oriol is also the first author.
이것은 저자가 확실히 알고 있으며, [ORDER MATTERS : SEQUENCE TO SEQUENCE FOR SETS](https://arxiv.org/pdf/1511.06391.pdf)에서 논의 된 것처럼 반복적인 주의 집중 메커니즘을 사용하여 해결할 수 있습니다. Oriol은 첫 번째 저자이기도 합니다.

I wish there was a comment on this point in the paper somewhere.
나는 논문에서 이 지점에 대한 의견이 어딘가에서 있었으면 좋겠다.

#### 2

이 접근 방식은 교육 사례 수가 늘어남에 따라 상당히 느려지지만, 일단이 수치가 커지면 매개 변수 방식으로 전환 할 수 있습니다.

#### 3

또한 교육 과정에서 특정 수의 예제 (예 : 5-25)를 사용한다는 점에 잠재적으로 관련이 있습니다. 따라서 이 숫자는 테스트 시간에도 사용해야합니다.

우리의 훈련 세트가 온라인으로 성장하기를 원한다면 어떻게 될까요? 교육 데이터의 인코더 LSTM이 더 많은 예제의 입력을 보는 데 "익숙하지"않기 때문에 네트워크를 재교육해야하는 것으로 보입니다. 반복적으로 학습 데이터를 서브 샘플링하거나, 여러 추론 패스 및 평균화 또는 이와 유사한 작업을 수행하지 않는 이상 그렇습니다.

우리가 FCE를 사용하지 않는다면 LSTM주의 메커니즘이 더 많은 예제를 다루는 데 여전히 "익숙"할 수는 없지만 이 문제의 중요성은 분명하지 않을 수 있습니다.

An interesting experiment would be to not use FCE and try to use 100 or 1000 training examples, while only training on up to 25 (with and fithout FCE).
흥미로운 실험은 FCE를 사용하지 않고 100 개 또는 1000 개의 교육 예제를 사용하는 것입니다. 또한 최대 25 개까지만 교육 할 수 있습니다 (FCE 포함 및 제외).

Discussion surrounding this point would be interesting.
이 점을 둘러싼 논의는 흥미로울 것이다.
